### ISSUE TITLE
htmlClass pattern validation is too restrictive

### ISSUE THREAD

**Issue by thomas-kl1** [magento/magento2#32004]
Resource: https://www.w3.org/TR/CSS22/syndata.html
Part 4.1.3

> In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+0080 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B\&W\?" or "B\26 W\3F".

### Preconditions (*)
1. Add a new container in your layout
2. Add a htmlClass with a character like "@"

### Steps to reproduce (*)
1. Add a new container in your layout
2. Add a htmlClass with a character like "@"

### Expected result (*)
1. The container is added to the dom with the correct class attribute value

### Actual result (*)
1. Layout validation fails because it says the character is not allowed

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by thomas-kl1** [comment]
@magento I am working on this

**Comment by ihor-sviziev** [comment]
Hi @thomas-kl1,
Could you describe what is the use case of using the `@` symbol in the class names? Magento definitely not supporting, but I don't think there is some valid use case for using such things.

**Comment by thomas-kl1** [comment]
@ihor-sviziev there is many css framework that use "@ " in classnames. Anyway the current validation it too much restrictive regarding this ressource https://www.w3.org/TR/CSS22/syndata.html

There is no valid point to prevent using other characters and some could be easily escaped thanks to the escaper object.
Many projects I work on, have to use patches to override this rule or to use alternative blocks and templates, which is redundant and could be avoided

**Comment by ihor-sviziev** [comment]
TBH I never saw that "at" symbol used in the class names, I also didn't see it in any CSS frameworks, and I personally don't think it's a good idea because it will confuse people. Maybe you could provide some examples?

I still believe this issue having a too-small amount of people. Probably priority for this issue should be P4, which is the lowest priority. If you have

**Comment by thomas-kl1** [comment]
You missed the point. The point: https://www.w3.org/TR/CSS21/syndata.html#value-def-identifier

If you need the full standard documentation, there's the links:

- https://www.w3.org/TR/CSS/#css
- https://www.w3.org/TR/css-syntax-3/#syntax-description
- https://www.w3.org/TR/css-syntax-3/#style-rule
- https://www.w3.org/TR/css-syntax-3/#parsing
- https://www.w3.org/TR/selectors-4/#class-html
- https://www.w3.org/TR/selectors-4/#id-selector

> I never saw that "at" symbol used in the class names, I also didn't see it in any CSS frameworks
That's your thought, it doesn't make it true. Some in-house and proprietary frameworks could use chars that are not part of the current rules: 

- `[a-zA-Z][a-zA-Z\d\-_:]*`
- `[a-zA-Z][a-zA-Z\d\-_]*(\s[a-zA-Z][a-zA-Z\d\-_]*)*`

> I personally don't think it's a good idea because it will confuse people
That's also your opinion, developers are free to use it or not, so there's no confusion either.

> I don't think there is some valid use case for using such things
Since some major companies are using it, it's pretty much valid to me, so it's still opinionated

What's the reason behind those Magento's rules? What is the legitimate design to write these regex pattern?
A framework should be a friendly tool for developers to help them to solve quickly and common problems. There it add a constraint without any reasons.

Also I've used the "@" character as an example, it could be any that follow the standard:

```
In CSS, identifiers (including element names, classes, and IDs in selectors) can contain only the characters [a-zA-Z0-9] and ISO 10646 characters U+00A0 and higher, plus the hyphen (-) and the underscore (_); they cannot start with a digit, two hyphens, or a hyphen followed by a digit. Identifiers can also contain escaped characters and any ISO 10646 character as a numeric code (see next item). For instance, the identifier "B&W?" may be written as "B\&W\?" or "B\26 W\3F".
```

We could agree or disagree to use such characters for identifiers, but we should follow the standards. The decision to use such characters is scoped to project, not to the framework.

**Comment by ihor-sviziev** [comment]
@sivaschenko @sidolov @gabrieldagama, could you review this issue and comment,s and provide your opinion?
Thank you!

**Comment by MeCapron** [comment]
Hello,

I'm reopening this issue because it is also a problem for us.

It makes even more sense now that Tailwind is around the place for a lot of project. The defined xsd is not scalable and do not take in account technologies or evolution.

As @thomas-kl1 mentioned, this is not only concerning the @ symbol but all symboles accepted by the W3 standards. 

Why can't we make this evolve?

**Comment by thomas-kl1** [comment]
at least colon is now accepted :') https://github.com/magento/magento2/commit/34115de04273a88f88ed070609e6316ede643617

**Comment by MeCapron** [comment]
Related to : https://github.com/magento/magento2/pull/36452 which may fix this issue

**Comment by engcom-November** [comment]
Hi @thomas-kl1 ,
Thanks for the report and collaboration! We have tried to reproduce this issue in Magento 2.4-develop branch but the issue is not reproducible.
We have followed the below steps:
1. Added a new container in your layout
<container name="some.container" as="someContainer" label="Some Container" htmlTag="div" htmlClass="some-container" />

Container added in app/code/Magento/Ui/view/base/templates/control/button/default.phtml

2.  Added a htmlClass with a character like "@"
We can able to see the container is added to the dom with the correct class attribute value
Kindly recheck the issue on Magento 2.4-develop instance and elaborate missing steps if the issue is still reproducible.
![image](https://user-images.githubusercontent.com/60198231/215080667-234ff540-0924-441e-9485-378482bc5744.png)

**Comment by thomas-kl1** [comment]
@engcom-November sure after 2 years without updates, maybe it have been fixed somewhere by someby...

**Comment by ihor-sviziev** [comment]
Hi @engcom-November,
From your screenshot I see you don't have attribute "class", but you have "htmlClass", I bet that's the reason why it work for you

**Comment by thomas-kl1** [comment]
Good catch @ihor-sviziev I didn't checked the screenshot 😅

**Comment by engcom-November** [comment]
Hello,
Hi @ihor-sviziev / @thomas-kl1 , Thank you for the response.
We have tried to reproduce the issue in Magento 2.4-develop branch and the issue has been reproducible for us. We have followed the below steps:
1. Open any front end file, for example considered /app/code/Magento/catalog/view/frontend/layout/catalog_product_view.xmlChange the 
2. html class name to htmlClass="w-screen left-1/2 right-1/2 mx-[-50vw] relative @something"
3. Deploy static content and flush cache
4. Front end - Open the Product details page from any catagory.
Were getting an error page
![image](https://user-images.githubusercontent.com/60198231/215679338-2404a6f2-4238-4a91-a22f-1139b3d14e25.png)

Hence confirming the issue.
